dali/operators/expressions/expression_impl_factory.h

```diff
@@ -21,80 +21,148 @@
+inline InputSamplePtr GetInputSamplePointer(DeviceWorkspace &ws, int input_idx, int sample_idx) {
+  return ws.template InputRef<GPUBackend>(input_idx).raw_tensor(sample_idx);
+}
+
+template <typename Backend>
+inline OutputSamplePtr GetOutput(const ExprFunc &func, workspace_t<Backend> &ws, TileDesc tile) {
+  return reinterpret_cast<char *>(GetOutputSamplePointer(ws, 0, tile.sample_idx)) +
+         tile.tile_size * tile.extent_idx * TypeTable::GetTypeInfo(func.GetTypeId()).size();
+}
+
+/**
+ * @brief Type erased obtaining pointers to inputs
+ */
+template <typename Backend>
+inline ArgPack GetArgPack(const ExprFunc &func, workspace_t<Backend> &ws,
+                          ConstantStorage<Backend> &st, const OpSpec &spec, TileDesc tile) {
+  ArgPack result;
+  result.resize(func.GetSubexpressionCount());
+  for (int i = 0; i < func.GetSubexpressionCount(); i++) {
+    if (func[i].GetNodeType() == NodeType::Function) {
+      DALI_FAIL("Function nodes not supported as subexpressions");
+    }
+    if (func[i].GetNodeType() == NodeType::Constant) {
+      const auto &constant = dynamic_cast<const ExprConstant &>(func[i]);
+      result[i] = st.GetPointer(constant.GetConstIndex(), constant.GetTypeId());
+    }
+    if (func[i].GetNodeType() == NodeType::Tensor) {
+      const auto &tensor = dynamic_cast<const ExprTensor &>(func[i]);
+      auto input_idx = tensor.GetInputIndex();
+      const auto *ptr =
+          reinterpret_cast<const char *>(GetInputSamplePointer(ws, input_idx, tile.sample_idx));
+      auto tile_offset =
+          tile.tile_size * tile.extent_idx * TypeTable::GetTypeInfo(func.GetTypeId()).size();
+      result[i] = ptr + tile_offset;
+    }
+  }
+  return result;
+}
+
+template <typename Backend>
+inline std::vector<ExtendedTileDesc> TransformDescs(const std::vector<TileDesc> &tiles,
+                                                    const ExprFunc &func, workspace_t<Backend> &ws,
+                                                    ConstantStorage<Backend> &st,
+                                                    const OpSpec &spec) {
+  std::vector<ExtendedTileDesc> result;
+  result.reserve(tiles.size());
+  for (auto &tile : tiles) {
+    result.push_back(
+        {tile, GetOutput<Backend>(func, ws, tile), GetArgPack(func, ws, st, spec, tile)});
+  }
+  return result;
+}
+
+template <template <ArithmeticOp, typename...> class ImplTensor, typename Backend>
+std::unique_ptr<ExprImplBase> ExprImplFactory1(const workspace_t<Backend> &ws,
+                                               const ExprFunc &expr) {
+  std::unique_ptr<ExprImplBase> result;
+  auto op = NameToOp(expr.GetFuncName());
+  auto input0_type = expr[0].GetTypeId();
+  TYPE_SWITCH(input0_type, type2id, Input0_t, ALLOWED_TYPES, (
+    VALUE_SWITCH(op, op_static, ALLOWED_UN_OPS, (
+      using Out_t = Input0_t;
+      if (expr[0].GetNodeType() == NodeType::Tensor) {
+        result.reset(new ImplTensor<op_static, Out_t, Input0_t>());
```

Comment:
Why not `make_unique`?

Reply:
Good point, done. :+1:

Reply:
Actually, it now compiles 3 times as long.

Reply:
:sob: